File: core/src/main/java/xtdb/vector/IRelationWriter.java
package xtdb.vector;

import clojure.lang.Keyword;
import org.apache.arrow.vector.types.pojo.Field;
import org.apache.arrow.vector.types.pojo.FieldType;

import java.util.Map;

@SuppressWarnings("try")
public interface IRelationWriter extends AutoCloseable, Iterable<Map.Entry<String, IVectorWriter>> {

    /**
     * <p>Maintains the next position to be written to.</p>
     *
     * <p>This is incremented either by using the {@link IRelationWriter#rowCopier}, or by explicitly calling {@link IRelationWriter#endRow()}</p>
     */
    IVectorPosition writerPosition();

    void startRow();
    void endRow();

    /**
     * This method syncs the value counts on the underlying writers/root (e.g. {@link org.apache.arrow.vector.VectorSchemaRoot#setRowCount})
     * so that all of the values written become visible through the Arrow Java API.
     * We don't call this after every write because (for composite vectors, and especially unions) it's not the cheapest call.
     */
    default void syncRowCount() {
        for (Map.Entry<String, IVectorWriter> entry : this) {
            entry.getValue().syncValueCount();
        }
    }

    IVectorWriter colWriter(String colName);

    @Deprecated
    IVectorWriter colWriter(Field field);
    IVectorWriter colWriter(String colName, FieldType fieldType);

    IRowCopier rowCopier(RelationReader relation);

    default void clear() {
        for (Map.Entry<String, IVectorWriter> entry : this) {
            entry.getValue().clear();
        }

        writerPosition().setPosition(0);
    }
}

File: core/src/main/kotlin/xtdb/vector/IRelationWriter.kt
package xtdb.vector

import org.apache.arrow.memory.BufferAllocator
import org.apache.arrow.vector.types.pojo.FieldType
import org.apache.arrow.vector.types.pojo.Schema
import xtdb.arrow.RelationReader
import xtdb.arrow.RelationWriter
import xtdb.arrow.RowCopier
import xtdb.arrow.VectorWriter

interface IRelationWriter : RelationWriter, AutoCloseable, Iterable<Map.Entry<String, VectorWriter>> {
    /**
     * Maintains the next position to be written to.
     *
     * This is incremented either by using the [IRelationWriter.rowCopier], or by explicitly calling [IRelationWriter.endRow]
     */
    override var rowCount: Int

    override fun rowCopier(rel: RelationReader): RowCopier {
        val copiers = rel.vectors.map {
            it.rowCopier(vectorFor(it.name, it.fieldType))
        }

        return RowCopier { srcIdx ->
            copiers.forEach { it.copyRow(srcIdx) }
            endRow()
        }
    }

    override fun openSlice(al: BufferAllocator) = asReader.openSlice(al)
    override fun openDirectSlice(al: BufferAllocator) = asReader.openDirectSlice(al)

    override val asReader get() = RelationReader.from(vectors.map { it.asReader }, rowCount)
}
